### PR TITLE
fix: support extended GeneralsOnline version format with build suffixes

### PR DESCRIPTION
**Pull request by undead2146**
## Summary

- **Root cause**: The Generals Online CDN changed the version format from `MMDDYY_QFE#` (2 segments) to `MMDDYY_QFE#_SUFFIX` (3+ segments, e.g. `"042826_QFE3_EAC"`). `ParseGeneralsOnlineVersion` required exactly 2 underscore-delimited parts, so it returned `null` for all new versions — update detection, version comparison, and sortable version IDs all silently broke.
- **Fix**: Changed the parser to accept 2+ segments, treating extra segments as build metadata that don't affect version ordering. Added a dedicated `CompareGeneralsOnlineVersions` method in `VersionComparer` that uses the structured parser instead of generic numeric comparison, ensuring `"042826_QFE2"` and `"042826_QFE2_EAC"` compare as equal.
- **Impact**: All 53 tests pass (existing + new). No changes to public API signatures — fully backward compatible with legacy 2-segment versions.

## Changes

| File | Change |
|---|---|
| `GameVersionHelper.cs` | `parts.Length != 2` → `parts.Length < 2` — accept extended format |
| `VersionComparer.cs` | New `CompareGeneralsOnlineVersions()` — structured date+QFE comparison with fallback |
| `GeneralsOnlineUpdateService.cs` | Updated comment to reflect extended format |
| `GeneralsOnlineApiResponse.cs` | Updated doc comment for version format |
| `GeneralsOnlineRelease.cs` | Updated doc comment for version format |
| `GeneralsOnlineConstants.cs` | Added remark clarifying separator usage |

## Tests added

| Test file | Coverage |
|---|---|
| `GameVersionHelperTests.cs` (new) | Legacy format, extended format, multiple suffixes, null/empty, invalid date, non-numeric QFE, sortable version equality across formats |
| `VersionComparerTests.cs` | Extended format theory: same version, higher QFE, legacy vs extended equality, newer date |
| `GeneralsOnlineJsonCatalogParserTests.cs` | manifest.json with extended version, latest.txt fallback with extended version |

## Test plan

- [x] All 53 version/comparison tests pass (20 new + 33 existing)
- [x] Build succeeds with 0 warnings, 0 errors
- [x] Backward compatible: legacy `MMDDYY_QFE#` versions still parse and compare correctly
- [x] Forward compatible: extended `MMDDYY_QFE#_SUFFIX` versions parse date+QFE correctly
- [x] Cross-format: legacy vs extended versions with same date+QFE compare as equal

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

## 🤖 OpenClaw Changes

**Files changed:** 2
**Lines added:** 7
**Lines removed:** 1
**Branch:** `fix/go-version-extended-format`
**Commit:** `c88bdda0`

*Last updated: 2026-05-04T09:14:14.969Z*

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes a version parsing regression caused by the Generals Online CDN switching from a 2-segment format (`MMDDYY_QFE#`) to a 3+-segment format (`MMDDYY_QFE#_SUFFIX`). The single-character guard change (`!= 2` → `< 2`) in `ParseGeneralsOnlineVersion` restores update detection, and the new `CompareGeneralsOnlineVersions` method in `VersionComparer` ensures structured date+QFE ordering is used instead of opaque numeric fallback.

- **`GameVersionHelper.cs`**: Relaxes the segment-count guard from exactly 2 to at least 2, and replaces the hardcoded `\"QFE\"` literal with `GeneralsOnlineConstants.QfeMarkerPrefix`, aligning with the project's constants convention.
- **`VersionComparer.cs`**: Introduces `CompareGeneralsOnlineVersions` with explicit handling for both-null (generic fallback), single-null (unparseable treated as older), and fully-parsed (date then QFE) cases — directly addressing the asymmetric fallback identified in a prior review.
- **Tests**: 20 new tests across three files cover legacy format, extended format, multi-suffix, null/empty, invalid inputs, mixed-parseable comparison, and catalog-parser integration.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the change is a minimal, well-tested relaxation of a length guard that restores broken version detection for the new CDN format.

The one-character logic fix is narrow and its correctness is directly validated by 20 new tests covering legacy, extended, multi-suffix, and edge-case inputs. The new CompareGeneralsOnlineVersions method explicitly handles all null-combination states, closing the asymmetric-fallback gap flagged in a prior review. No public API signatures changed, backward compatibility with existing 2-segment versions is confirmed, and the previously-hardcoded QFE literal is now correctly referenced through the constants class.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| GenHub/GenHub.Core/Helpers/GameVersionHelper.cs | Core fix: relaxes parts.Length != 2 to < 2, uses QfeMarkerPrefix constant instead of hardcoded QFE, and updates doc comments for extended format support |
| GenHub/GenHub.Core/Helpers/VersionComparer.cs | Adds CompareGeneralsOnlineVersions() using structured date+QFE parsing; correctly handles both-null (fallback to numeric), single-null (unparseable treated as older), and fully-parsed cases; braces-only style cleanup elsewhere |
| GenHub/GenHub.Tests/GenHub.Tests.Core/Helpers/GameVersionHelperTests.cs | New test file with comprehensive coverage of legacy, extended, multi-suffix, null/empty, invalid date, non-numeric QFE, and sortable version equality cases |
| GenHub/GenHub.Tests/GenHub.Tests.Core/Helpers/VersionComparerTests.cs | New theory-based tests for extended format comparison, mixed-parseable fallback, and both-unparseable fallback; all assertions use Math.Sign correctly |
| GenHub/GenHub/Features/Content/Services/GeneralsOnline/GeneralsOnlineManifestFactory.cs | Comment-only changes: adds tracking context for filesWithHashes and corrects a stale comment that incorrectly described the game-client manifest as including map files |

</details>

</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["Version string input\ne.g. 042826_QFE3_EAC"] --> B["Split on underscore\nRemoveEmptyEntries"]
    B --> C{parts.Length less than 2?}
    C -- Yes --> D["Return null"]
    C -- No --> E["datePart = parts 0\nqfePart = parts 1 minus QFE prefix"]
    E --> F{Valid 6-char date\nand numeric QFE?}
    F -- No --> D
    F -- Yes --> G["Return DateTime + Qfe\nExtra suffix parts ignored"]

    G --> H["CompareGeneralsOnlineVersions"]
    H --> I{Parsed results?}
    I -- Both null --> J["CompareNumericVersions fallback"]
    I -- One null --> K["Null treated as older"]
    I -- Both parsed --> L["Compare Date then QFE"]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["Version string input\ne.g. 042826_QFE3_EAC"] --> B["Split on underscore\nRemoveEmptyEntries"]
    B --> C{parts.Length less than 2?}
    C -- Yes --> D["Return null"]
    C -- No --> E["datePart = parts 0\nqfePart = parts 1 minus QFE prefix"]
    E --> F{Valid 6-char date\nand numeric QFE?}
    F -- No --> D
    F -- Yes --> G["Return DateTime + Qfe\nExtra suffix parts ignored"]

    G --> H["CompareGeneralsOnlineVersions"]
    H --> I{Parsed results?}
    I -- Both null --> J["CompareNumericVersions fallback"]
    I -- One null --> K["Null treated as older"]
    I -- Both parsed --> L["Compare Date then QFE"]
```

</a>
</details>

<sub>Reviews (7): Last reviewed commit: ["OpenClaw: Address review feedback from @..."](https://github.com/community-outpost/genhub/commit/b981f10eee562cc42802db170f1beb264eaee8ea) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=30537655)</sub>

### CODE REVIEW

**Comment by undead2146** [comment]
🤖 **OpenClaw task queued**

Task ID: `openclaw-1777688912853`

Another task is currently running on this PR. Your task will be processed when the current task completes.

Commands:
- Address code review feedback from @coderabbitai[bot]:

1. GenHub/GenHub.Core/Helpers/VersionComparer.cs:77 - <a href="#">P2</a> **Asymmetric fallback when only one version is parseable**  When exactly one of the t

**Comment by undead2146** [comment]
🤖 **OpenClaw is processing your request...**

Task ID: `openclaw-1777689090744`

Commands:
- Address code review feedback from @coderabbitai[bot]:

1. GenHub/GenHub.Core/Helpers/VersionComparer.cs:77 - <a href="#">P2</a> **Asymmetric fallback when only one version is parseable**  When exactly one of the t

**Comment by ElTioRata** [comment]
Updater works again. EAC startup is still needed to play online.
imagen

**Comment by undead2146** [comment]
@undead2146 openclaw: Address this comment (ID: 4362663878): "


<details>
<summary>📝 Walkthrough</summary>

## Walkthrough

Extended Generals Online version string ..."

**Comment by undead2146** [comment]
🤖 **OpenClaw task queued**

Task ID: `openclaw-1777862795000`

Another task is currently running on this PR. Your task will be processed when the current task completes.

Commands:
- Address this comment (ID: 4362676789): "

**Comment by undead2146** [comment]
> Updater works again. EAC startup is still needed to play online. imagen

Does the user have to run EAC before running GO?

File: GenHub/GenHub.Core/Helpers/VersionComparer.cs [hunk 1]
```diff
@@ -41,14 +41,42 @@ public static int CompareVersions(string? version1, string? version2, string? pu
...
+            if (dateCompare != 0)
+            {
+                return dateCompare;
+            }
+
+            return parsed1.Value.Qfe.CompareTo(parsed2.Value.Qfe);
+        }
+
+        // Fallback for unparseable versions (e.g. "Unknown", semantic versions)
+        return CompareNumericVersions(version1, version2);
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub.Core/Helpers/VersionComparer.cs]: ✅ Addressed by OpenClaw (Task: `openclaw-1777885675931`)

File: GenHub/GenHub/Features/Content/Services/GeneralsOnline/GeneralsOnlineManifestFactory.cs [hunk 2]
```diff
@@ -408,12 +408,16 @@ private async Task<List<ContentManifest>> UpdateManifestsWithExtractedFiles(
...
-        List<(string RelativePath, FileInfo FileInfo, string Hash, bool IsMap)> filesWithHashes = [];
+        List<(string RelativePath, FileInfo FileInfo, string Hash, bool IsMap, bool IsPlugin)> filesWithHashes = [];
 
         // Detect Maps directory (case-insensitive)
         var mapsDirectory = Directory.GetDirectories(extractPath, "*", SearchOption.TopDirectoryOnly)
             .FirstOrDefault(d => Path.GetFileName(d).Equals(GeneralsOnlineConstants.MapsSubdirectory, StringComparison.OrdinalIgnoreCase));
 
+        // Detect plugins directory (case-insensitive) for EAC and other plugins
+        var pluginsDirectory = Directory.GetDirectories(extractPath, "*", SearchOption.TopDirectoryOnly)
+            .FirstOrDefault(d => Path.GetFileName(d).Equals("plugins", StringComparison.OrdinalIgnoreCase));
```
[the remark this answers is not in the document]
Comment (undead2146) [reply GenHub/GenHub/Features/Content/Services/GeneralsOnline/GeneralsOnlineManifestFactory.cs]: ✅ Addressed by OpenClaw (Task: `openclaw-1777885675931`)

**Comment by bobtista** [comment]
I'd move the EAC plugin work to another PR, make sure there is a real behavioral change, add tests, and try it to make sure it actually works. 
nit: inconsistent usage of braces with single line if () {} blocks
Update the PR description too
Otherwise, I think the version-format fix looks solid

**Comment by bobtista** [comment]
Closing as superseded by #297, which is now merged into `development` (b7b6a0c).

#297 covers this PR's fix and generalises it:

- `MmddyyQfeVersionScheme` parses `MMDDYY`, the `QFE` revision, and any number of trailing build tags — `060526_QFE1`, `042826_QFE3_EAC` and `011526_QFE1_EAC_X86` all parse, and the trailing tags are deliberately excluded from ordering. That is the extended-format behaviour this PR added.
- Ordering is done on a structured `ContentVersion` (year, month, day, QFE) rather than on `MMDDYY` as an integer, so it also fixes cross-year ordering (`060526_QFE1` after `121525_QFE1`) that this PR's date-then-QFE comparison did not address on its own.
- #297 removes `ParseGeneralsOnlineVersion` entirely, replacing it with `GetGeneralsOnlineManifestIdComponent` for manifest-ID encoding only. The lines this PR changes no longer exist on `development`, so it can no longer merge cleanly.

Tracked in #307. Thanks for the fix — the parsing behaviour it identified is preserved.

**Comment by undead2146** [comment]
🤖 **Auto-Response Triggered**

Address this comment (ID: 5117778741): "Closing as superseded by #297, which is now merged into `development` (b7b6a0c).

#297 covers this PR's fix and generalises it:

- `MmddyyQfeVersionScheme` parses `MMDDYY`, the `QFE` revision, and any..."

*This was auto-generated by the OpenClaw auto-response monitor.*
